tiled/_tests/test_authentication.py

```diff
@@ -449,3 +449,151 @@ def test_sticky_identity(enter_password, config):
     # Clear the default.
     clear_default_identity(context.api_uri)
     get_default_identity(context.api_uri) is None
+
+
+@pytest.fixture
+def principals_context(enter_password, config):
+    """
+    Fetch UUID for an admin and an ordinary user; include the client context.
+    """
+    # Make alice an admin. Leave bob as a user.
+    config["authentication"]["tiled_admins"] = [{"provider": "toy", "id": "alice"}]
+
+    with Context.from_app(build_app_from_config(config)) as context:
+        # Log in as Alice and retrieve admin UUID for later use
+        with enter_password("secret1"):
+            context.authenticate(username="alice")
+
+        principal = context.whoami()
+        assert "admin" in (role["name"] for role in principal["roles"])
+        admin_uuid = principal["uuid"]
+        context.logout()
+
+        # Log in as Bob and retrieve Bob's UUID for later use
+        with enter_password("secret2"):
+            context.authenticate(username="bob")
+
+        principal = context.whoami()
+        assert "admin" not in (role["name"] for role in principal["roles"])
+        bob_uuid = principal["uuid"]
+        context.logout()
+
+        yield {
+            "uuid": {"alice": admin_uuid, "bob": bob_uuid},
+            "context": context,
+        }
+
+
+@pytest.mark.parametrize(
+    "username, scopes, resource",
+    (
+        ("alice", ["read:principals"], "/api/v1/auth/principal"),
+        ("bob", ["read:data"], "/api/v1/array/full/A1"),
+    ),
+)
+def test_admin_api_key_any_principal(
+    enter_password, principals_context, username, scopes, resource
+):
+    """
+    Admin can create usable API keys for any prinicipal, within that principal's scopes.
+    """
+    with principals_context["context"] as context:
+        # Log in as Alice, create and use API key after logout
+        with enter_password("secret1"):
+            context.authenticate(username="alice")
+
+        principal_uuid = principals_context["uuid"][username]
+        api_key = _create_api_key_other_principal(
+            context=context, uuid=principal_uuid, scopes=scopes
+        )
+        assert api_key
+        context.logout()
+
+        url_with_key = f"{resource}?api_key={api_key}"
```

Comment:
The API key can be placed in three places:

1. Query parameter `?api_key=SECRET`
2. Header as `Authorization:Apikey SECRET`
3. Cookie `tiled_api_key=SECRET`

The query parameter method is convenient because it can be shared by copy/paste, but it's also less secure. (For one, URLs tend to be logged. For another, if it is used in a browser, JavaScript on the page has access to the URL.) There have been some discussions about making support for API keys in query parameters configurable, and maybe even off by default, for security.

With that in mind, for future proofing, use

```py
context.api_key = api_key
```

which will set the `Authorization` header:

https://github.com/bluesky/tiled/blob/204676bbb841d252c1413abfe2646c3b7d27a525/tiled/client/context.py#L335-L341

and then you can simple use `context.http_client.get(resource)` below.

Reply:
That's helpful. Thanks!